src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs

```diff
@@ -0,0 +1,149 @@
+        ///     by TelemetryState.ObservedRuleChanges
+        ///  3. Telemetry fires with 'Resolved' label when 1. and 2. are no longer true
+        ///  4. Telemetry stops firing once all target frameworks observe an Evaluation following
+        ///     the first design time build, or once it has fired once with 'Resolved' label. 
+        ///     This prevents telemetry events caused by project changes and tree updates after initial load has completed.
+        /// </summary>
+        public void ObserveTreeUpdateCompleted()
+        {
+            if (stopTelemetry) return;
+
+            stopTelemetry = isReadyToResolve || IsStopTelemetryInAllTargetFrameworks();
+
+            _telemetryService.PostProperty($"{TelemetryEventName}/{(isReadyToResolve ? ResolvedLabel : UnresolvedLabel)}",
+                ProjectProperty, projectId);
+        }
+
+        public void ObserveUnresolvedRules(ITargetFramework targetFramework, IEnumerable<string> rules)
+        {
+            if (stopTelemetry) return;
+
+            var telemetryState = telemetryStates.GetOrAdd(targetFramework, (key) => new TelemetryState());
+
+            foreach (var rule in rules)
+            {
+                // observe all unresolved rules by default ignoring whether they have actual changes
+                UpdateObservedRuleChanges(telemetryState, rule);
+            }
+        }
+
+        public void ObserveHandlerRulesChanges(
+            ITargetFramework targetFramework,
+            IEnumerable<string> handlerRules, 
+            IImmutableDictionary<string, IProjectChangeDescription> projectChanges)
+        {
+            if (stopTelemetry) return;
+
+            if (telemetryStates.TryGetValue(targetFramework, out var telemetryState))
+            {
+                foreach (var rule in handlerRules)
+                {
+                    var hasChanges = projectChanges.TryGetValue(rule, out var description)
+                        && description.Difference.AnyChanges;
+                    UpdateObservedRuleChanges(telemetryState, rule, hasChanges);
+                }
+            }
+        }
+
+        public void ObserveCompleteHandlers(ITargetFramework targetFramework, RuleHandlerType handlerType)
+        {
+            if (stopTelemetry) return;
+
+            if (telemetryStates.TryGetValue(targetFramework, out var telemetryState))
+            {
+                telemetryState.ObservedDesignTime |= handlerType == RuleHandlerType.DesignTimeBuild;
+                telemetryState.StopTelemetry |= telemetryState.ObservedDesignTime && handlerType == RuleHandlerType.Evaluation;
+            }
+
+            isReadyToResolve = !telemetryStates.IsEmpty && telemetryStates.Values.All(s => s.IsReadyToResolve());
+
+            // if isReadyToResolve, wait until atleast one Resolved telemetry event has fired before stopping telemetry
+            stopTelemetry = !isReadyToResolve && IsStopTelemetryInAllTargetFrameworks();
```

Comment:
>stopTelemetry [](start = 12, length = 13)

Race condition here. if two threads simultaneously run ObserveCompleteHandlers and ObserveTreeUpdateCompleted there is a chance that ObserveTreeUpdateCompleted would set stopTelemetry to true when isReadyToResolve=false, but ObserveCompleteHandlers  would set it immediatelly back to false. (when threads run simultaneously asuume you are already past the point where you check: if (stopTelemetry) return;

Reply:
Also we do know that bool is atomic, but what about multiple CPU and their caches? You don't seem to synchronize bool flags at all in this class?

---
In reply to: [145180963](https://github.com/dotnet/project-system/pull/2879#discussion_r145180963) [](ancestors = 145180963)